src/raw_rtde/connection.rs

```diff
@@ -0,0 +1,891 @@
+    fn send_robot_input(
+        &mut self,
+        io_setup: &IoSetup,
+        input_data: &[RTDEData],
+    ) -> Result<(), String>;
+
+    /// Waits for a packet send by the robot.
+    ///
+    /// One needs to call `configure_outputs` and `start_robot_to_host_transmition` to receive such packets.
+    ///
+    /// May return None is someone used the waker from `get_wait_for_robot_output_waker` to interrupt the call.
+    fn wait_for_robot_output(
+        &mut self,
+        io_setup: &IoSetup,
+    ) -> Result<Option<Vec<RTDEData>>, String>;
+
+    /// Returns a closure that can be called to interrupt a wait_for_robot_output (happening in another thread).
+    fn get_wait_for_robot_output_interrupter(
+        &self,
+    ) -> Box<dyn Fn() -> Result<(), String> + Send + 'static>;
+}
+
+impl GenericRawRTDE<TcpStream> {
+    /// Connect to a given TCP/IP port and negotiate the protocol version.
+    pub fn connect(addr: SocketAddr) -> Result<Self, String> {
+        let stream =
+            TcpStream::connect(addr).map_err(|e| format!("Error connecting to RTDE: {}", e))?;
+
+        let mut result = GenericRawRTDE::new(stream)?;
+        result.negotiate_protocol_version()?;
+        Ok(result)
+    }
+}
+
+impl<T: Read + Write + event::Source> GenericRawRTDE<T> {
+    /// Create a new connection for a given stream.
+    ///
+    /// This is a low-level call (for most purposes one should use `RawRTDE::connect`) and protocol version
+    /// negotation is not performed.
+    pub fn new(stream: T) -> Result<Self, String> {
+        let poll =
+            Poll::new().map_err(|e| format!("Error starting RTDE - when create a Poll: {e}"))?;
+
+        let waker = Arc::new(
+            Waker::new(poll.registry(), WAKE_TOKEN)
+                .map_err(|e| format!("Error creating waker: {e}"))?,
+        );
+
+        Ok(GenericRawRTDE {
+            stream,
+            poll,
+            waker,
+            wait_for_reply_timeout: RTDE_WAIT_FOR_REPLY_TIMEOUT,
+        })
+    }
+
+    /// Asks the control about its version.
+    pub fn get_controller_version(&mut self) -> Result<ControllerVersion, String> {
+        let request = Self::encode_controller_version_request()?;
+        self.send_packet(&request)?;
+        let reply = self.wait_for_reply(RTDECommand::GetControllerVersion as u8)?;
+        Self::decode_controller_version_reply(reply)
+    }
+
+    /// Negotiates the protocol version.
+    ///
+    /// Needs to be done before sending other packets. `RawRTDE::connect` does it automatically.
+    pub fn negotiate_protocol_version(&mut self) -> Result<(), String> {
+        let request = Self::encode_protocol_version_request(RTDE_VERSION)?;
+        self.send_packet(&request)?;
+        let reply = self.wait_for_reply(RTDECommand::RequestProtocolVersion as u8)?;
+        let protocol_ok = Self::decode_protocol_version_reply(reply)?;
+
+        match protocol_ok {
```

Comment:
`match ... { true => ..., false => ... }` can and should be simplified to `if-else`.